src/aof.zig

```diff
@@ -0,0 +1,655 @@
+        errdefer os.close(dir_fd);
+
+        const basename = std.fs.path.basename(absolute_path);
+
+        return AOF.init(dir_fd, basename);
+    }
+
+    fn init(dir_fd: os.fd_t, relative_path: []const u8) !AOF {
+        const fd = try IO.open_file(dir_fd, relative_path, 0, true);
+        try os.lseek_END(fd, 0);
+
+        return AOF{ .fd = fd };
+    }
+
+    pub fn close(self: *AOF) void {
+        os.close(self.fd);
+    }
+
+    pub fn prepare_entry(last_checksum: *?u128, message: *const Message, options: struct { replica: u64, primary: u64 }, entry: *AOFEntry) void {
+        const body = message.body();
+
+        assert(body.len < constants.message_size_max);
+
+        // When writing, entries can backtrack / duplicate, so we don't necessarily have a valid chain.
+        // Still, log when that happens. The `aof prepare` command can generate a consistent file from
+        // entries like these.
+        log.debug("{}: aof prepare_entry: parent {} (should == {}) our checksum {}", .{ options.replica, message.header.parent, last_checksum.*, message.header.checksum });
+        if (last_checksum.* == null or last_checksum.*.? != message.header.parent) {
+            log.info("{}: aof prepare_entry: parent {}, expected {} instead", .{ options.replica, message.header.parent, last_checksum.* });
+        }
+        last_checksum.* = message.header.checksum;
+
+        // cluster is in the VSR header so we don't need to store it explcitly.
+        entry.* = AOFEntry{
+            .metadata = AOFMetadata{ .vsr_header = message.header.*, .replica = options.replica, .primary = options.primary },
+            .body_size = body.len,
+            .body = undefined,
+        };
+
+        stdx.copy_disjoint(.exact, u8, entry.body[0..body.len], body);
+        std.mem.set(u8, entry.body[body.len..constants.message_size_max], 0);
+    }
+
+    /// Write a message to disk. Once this function returns, the data passed in
+    /// is guaranteed to have been written using O_DIRECT and O_SYNC and
+    /// can be considered safely persisted for recovery purposes once this
+    /// call returns.
+    ///
+    /// We purposefully use standard disk IO here, and not IO_uring. It'll
+    /// be slower and have syscall overhead, but it's considerably more
+    /// battle tested.
+    ///
+    /// We don't bother returning a count of how much we wrote. Not being
+    /// able to fully write the entire payload is an error, not an expected
+    /// condition.
+    pub fn write(self: *AOF, message: *const Message, options: struct { replica: u64, primary: u64 }) !void {
+        var entry: AOFEntry align(constants.sector_size) = undefined;
+        prepare_entry(&self.last_checksum, message, .{ .replica = options.replica, .primary = options.primary }, &entry);
+
+        const padded_size = entry.calculate_padded_size();
+
+        // writeAll logic - in case we're interrupted by a signal. Needed for Direct IO? Alignment?
```

Comment:
More an open question - can a Direct IO write be interrupted by a signal? If so, when resuming, does anything need to be done for alignment (if you know offhand - otherwise I'll do a bit more digging!)

Reply:
I have no idea :)

Reply:
Did a bit of digging here...
`man signal`:
> Note that a (local) disk is not a slow device according to this definition; I/O operations on disk devices are not interrupted by signals.

and [here](https://stackoverflow.com/questions/61719156/posix-partial-write-and-signal-interrupts):
> Regarding the note, if write is returning the full nbytes after a signal arriving, the signal handler was non-interrupting. This is the default on Linux with any modern libc (glibc, musl, anything but libc5 basically), and it's almost always the right thing. If you actually want interrupting signals you have to install the signal handler with sigaction and the SA_RESTART flag clear. (And conversely if you're installing signal handlers you want to have the normal, reasonable, non-interrupting behavior, for portability you should use sigaction and set the SA_RESTART flag rather than using the legacy function signal).

On top of that... We don't actually have any signal handlers; but I wrote a quick test to confirm all this:

```zig
const std = @import("std");
const os = std.os;
const assert = std.debug.assert;

pub fn handler(param: c_int) callconv(.C) void {
    _ = param;
    std.log.info("hello from signal", .{});
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    const buffer = try allocator.allocAdvanced(u8, 4096, 10 * 1024 * 1024 * 1024, .exact);
    const disk_size = buffer.len;

    var flags: u32 = os.O.CLOEXEC | os.O.RDWR | os.O.DSYNC;
    var mode: os.mode_t = 0;

    if (@hasDecl(os.O, "LARGEFILE")) flags |= os.O.LARGEFILE;

    flags |= os.O.DIRECT;
    flags |= os.O.CREAT;
    mode = 0o666;

    assert((flags & os.O.DSYNC) > 0);

    const dir_fd = try os.open(".", os.O.CLOEXEC | os.O.RDONLY, 0);
    const fd = try os.openat(dir_fd, "test.file", flags, mode);

    try os.fsync(fd);

    const act = std.os.linux.Sigaction{
        .handler = .{ .handler = handler },
        .mask = os.empty_sigset,
        .flags = 0,
    };
    _ = std.c.sigaction(2, &act, null);

    std.log.info("hit ctrl-c when ready", .{});

    var index: usize = 0;
    while (index != disk_size) {
        std.log.info("looping", .{});
        index += try os.write(fd, buffer[index..disk_size]);
    }
}
```

The `writeAll` logic is still needed for huge writes (like in this test) but hitting ctrl-c doesn't interrupt the syscall.